federated_learning/libs/datasets/toxicity.py

```diff
@@ -1,11 +1,147 @@
 #!/usr/bin/env python
 # coding: utf-8
+import collections
+import os
 
+import pandas as pd
+import numpy as np
+import tensorflow as tf
+import tensorflow_federated as tff
+from tensorflow_federated.python.simulation import ClientData
 
-def load_data():
-    pass
+from rdkit import Chem
+from rdkit.Chem import rdmolops
+from rdkit.Chem import SaltRemover
+from ..utils import create_ids, one_hot, pad_bottom_right_matrix, pad_bottom_matrix
 
-class ToxicityDataset:
-    # use tox_AMES_summary.20201106
-    pass
 
+def load_data(FL_FLAG, dataset_name, max_n_atoms, max_n_types, n_groups=2, subset_ratios: list = None):
+    if FL_FLAG:
+        dataset = ToxicityDataset(
+            dataset_name, max_n_atoms, max_n_types, n_groups, subset_ratios)
+    else:
+        dataset = create_dateset(dataset_name, max_n_atoms, max_n_types)
+    return dataset
+
+
+def _read_dataset_file(dataset_name: str):
+    file_path = os.path.join('data', 'tox_AMES_summary.20201106.xlsx')
+    return pd.read_excel(file_path, sheet_name=dataset_name)
+
+
+def _create_mol_feature(mol, max_n_types):
+    """
+    creates a feature vector from atoms in a molecule
+
+    mol: `Chem.rdchem.Mol`
+    """
+    mol_features = np.array([atom.GetAtomicNum() for atom in mol.GetAtoms()])
+    mol_features = one_hot(mol_features, max_n_types).astype(np.int32)
+    return mol_features
+
+
+def _create_element(data_row, max_n_atoms, max_n_types, salt_remover):
+    """
+    creates an element for `tf.Dataset`
+
+    data_row: an element of an iterator obtained by `data.itertuples` 
+    """
+    label = np.array(1 if data_row.Label == "Positive" else 0, dtype=np.int32)
+    mol = Chem.MolFromSmiles(data_row.Canonical_SMILES)
+    salt_removed_mol = salt_remover.StripMol(mol, dontRemoveEverything=True)
+    features = pad_bottom_matrix(
+        _create_mol_feature(salt_removed_mol, max_n_types), max_n_atoms)
+    adjs = pad_bottom_right_matrix(
+        rdmolops.GetAdjacencyMatrix(salt_removed_mol), max_n_atoms)
+    return ({'adjs': adjs, 'features': features}, label)
```

Comment:
Moved above two methods out of class, because these methods are also used in `create_dateset`.

Reply:
I see. Thank for refactoring.